### PR TITLE
Added option to merge launches

### PR DESCRIPTION
**Pull request by VladislavAkulich**

### CODE REVIEW

File: lib/helpers.js [hunk 1]
```diff
@@ -19,4 +22,17 @@ module.exports = {
...
     },
 
+    readLaunchesFromFile() {
+        return fs.readFileSync(tempFileForLaunchIds, 'utf8', function(err, data) {
+            if (err) throw err;
+            return data;
+        }).toString().split('|');
+    },
+
+    saveLaunchIdToFile(launch_id) {
```
Comment (avarabyeu) [inline lib/helpers.js]: can we make the file in the same format as we have in jvm-based clients?
https://github.com/reportportal/client-java/blob/master/src/main/java/com/epam/reportportal/utils/LaunchFile.java#L41

**Review by avarabyeu — CHANGES_REQUESTED** [review]

File: lib/report-portal-client.js [hunk 2]
```diff
@@ -138,6 +138,7 @@ class RPClient {
                 this.restClient.create(url, launchData, { headers: this.headers })
                     .then((response) => {
                         this.map[tempId].realId = response.id;
+                        helpers.saveLaunchIdToFile(responce.id);
```
Comment (BorisOsipov) [inline lib/report-portal-client.js]: Make option for this. I don't need such files in my test project workspace.

File: lib/report-portal-client.js [hunk 3]
```diff
@@ -199,6 +200,38 @@ class RPClient {
         };
     }
 
+    /*
+     * @Returns {Object} - an object which contains a data for merge launches in ReportPortal.
+     */
+    getMergeLaunchesData(launchIds) {
```
Comment (BorisOsipov) [inline lib/report-portal-client.js:213]: missing jsdoc

File: lib/report-portal-client.js [hunk 4]
```diff
@@ -199,6 +200,38 @@ class RPClient {
...
+            "name": this.config.launch || 'Test launch name',
+            "tags": this.config.tags
+        };
+    }
+
+    /*
+     * This method is used for merge launches in ReportPortal.
+     */
+    mergeLaunches() {
+        let url = [this.baseURL, "launch", "merge"].join("/");
```
Comment (BorisOsipov) [inline lib/report-portal-client.js]: let => const

File: lib/report-portal-client.js [hunk 5]
```diff
@@ -199,6 +200,38 @@ class RPClient {
...
+
+        let launchIds = helpers.readLaunchesFromFile();
```
Comment (BorisOsipov) [inline lib/report-portal-client.js]: let => const

File: lib/report-portal-client.js [hunk 6]
```diff
@@ -199,6 +200,38 @@ class RPClient {
...
+        launchIds.pop();
+        let data = this.getMergeLaunchesData(launchIds);
```
Comment (BorisOsipov) [inline lib/report-portal-client.js]: let => const

File: lib/helpers.js [hunk 7]
```diff
@@ -19,4 +22,17 @@ module.exports = {
             .request(method, url, request, options);
     },
 
+    readLaunchesFromFile() {
+        return fs.readFileSync(tempFileForLaunchIds, 'utf8', function(err, data) {
```
Comment (BorisOsipov) [inline lib/helpers.js]: You can't pass callback for `fs.readFileSync`. you have to try-catch exceptions

File: lib/helpers.js [hunk 8]
```diff
@@ -19,4 +22,17 @@ module.exports = {
             .request(method, url, request, options);
     },
 
+    readLaunchesFromFile() {
```
Comment (BorisOsipov) [inline lib/helpers.js:33]: will be nice to have test for both readLaunchesFromFile\saveLaunchIdToFile

File: .nyc_output/3feadddc-a951-4e57-a7b8-4217232779ca.json [hunk 9]
```diff
@@ -0,0 +1 @@
...39,"column":32},"end":{"line":86,"column":5}},"line":39},"6":{"name":"(anonymous_6)","decl":{"start":{"line":40,"column":48},"end":{"line":40,"column":49}},"loc":{"start":{"line":40,"column":58},"end":{"line":57,"column":9}},"line":40},"7":{"name":"(anonymous_7)","decl":{"start":{"line":51,"column":57},"end":{"line":51,"column":58}},"loc":{"start":{"line":51,"column":69},"end":{"line":56,"column":13}},"line":51},"8":{"name":"(anonymous_8)","decl":{"start":{"line":59,"column":37},"end":{"line":59,"column":38}},"loc":{"start":{"line":59,"column":47},"end":{"line":71,"column":9}},"line":59},"9":{"name":"(anonymous_9)","decl":{"start":{"line":64,"column":58},"end":{"line":64,"column":59}},"loc":{"start":{"line":64,"column":69},"end":{"line":70,"column":13}},"line":64},"10":{"name":"(anonymous_10)","decl":{"start":{"line":73,"column":33},"end":{"line":73,"column":34}},"loc":{"start":{"line":73,"column":43},"end":{"line":85,"column":9}},"line":73},"11":{"name":"(anonymous_11)","decl":{"start":{"line":78,"column":58},"end":{"line":78,"column":59}},"loc":{"start":{"line":78,"column":69},"end":{"line":84,"column":13}},"line":78},"12":{"name":"(anonymous_12)","decl":{"start":{"line":88,"colu
```
Comment (BorisOsipov) [inline .nyc_output/3feadddc-a951-4e57-a7b8-4217232779ca.json]: Remove this file and add ` .nyc_output/` to gitignore

**Review by BorisOsipov — CHANGES_REQUESTED** [review]

File: lib/report-portal-client.js (hunk 5 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: I don't see any purpose to make it const. It is just variable, not constant. Moreover, I change it in the next line.

File: lib/report-portal-client.js (hunk 4 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: I don't see any purpose to make it const. It is just variable, not constant.

File: lib/report-portal-client.js (hunk 6 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: I don't see any purpose to make it const. It is just variable, not constant.

File: .nyc_output/3feadddc-a951-4e57-a7b8-4217232779ca.json (hunk 9 above)
Comment (VladislavAkulich) [reply .nyc_output/3feadddc-a951-4e57-a7b8-4217232779ca.json]: Done

File: lib/helpers.js (hunk 7 above)
Comment (VladislavAkulich) [reply lib/helpers.js]: Thanks, good point! Done

File: lib/report-portal-client.js (hunk 3 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js:213]: Done

File: lib/report-portal-client.js (hunk 2 above)
Comment (BorisOsipov) [reply lib/report-portal-client.js]: also `responce` => `response`

File: lib/helpers.js (hunk 1 above)
Comment (VladislavAkulich) [reply lib/helpers.js]: I've added the same behavior for JS Client.

File: lib/report-portal-client.js (hunk 2 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: Done

File: lib/helpers.js (hunk 8 above)
Comment (VladislavAkulich) [reply lib/helpers.js:33]: I am going to add tests for these functions a bit later. Currently I don't have enough time due to sprint issues.

File: lib/report-portal-client.js (hunk 4 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: Linter did it

File: lib/report-portal-client.js (hunk 5 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: Linter did it

File: lib/report-portal-client.js (hunk 6 above)
Comment (VladislavAkulich) [reply lib/report-portal-client.js]: Linter did it

**Review by avarabyeu — APPROVED** [review]

**Comment by BorisOsipov** [comment]
@avarabyeu @DzmitryHumianiuk It is cool to merge code without UT... It will definitely make code base better and attract new contributors. Keep going the same way and 
don't forget repeat complaning "we don't have js contributors"

**Comment by DzmitryHumianiuk** [comment]
@BorisOsipov what stands for `UT`? :)

and thank you for code review!

**Comment by BorisOsipov** [comment]
@DzmitryHumianiuk unit tests

**Comment by DzmitryHumianiuk** [comment]
@BorisOsipov these enhancements made as a part of client offering. Literally pre-sale in firefighting mode.
Guys will back to it later. Hopefully.

let's keep an eye on it.

**Comment by BorisOsipov** [comment]
>these enhancements made as a part of client offering. Literally pre-sale in firefighting mode.

mmm it isn't reason to merge any code to master. for pre-sale\prototype you can use specific commit from any branch with any experimental changes e.g.
```
"dependencies": {
...
"reportportal-client": "https://github.com/reportportal/client-javascript.git#b09e49b"
...
}
```

¯\_(ツ)_/¯

**Pull request merged**